runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerContainerEnvironment.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.fnexecution.environment;
+
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.beam.model.pipeline.v1.RunnerApi.Environment;
+import org.apache.beam.runners.fnexecution.control.InstructionRequestHandler;
+
+/**
+ * A {@link RemoteEnvironment} that wraps a running Docker container.
+ *
+ * <p>A {@link DockerContainerEnvironment} owns both the underlying docker container that it
+ * communicates with an the {@link InstructionRequestHandler} that it uses to do so.
+ */
+@ThreadSafe
+class DockerContainerEnvironment implements RemoteEnvironment {
+
+  static DockerContainerEnvironment create(
+      DockerCommand docker,
+      Environment environment,
+      String containerId,
+      InstructionRequestHandler instructionHandler) {
+    return new DockerContainerEnvironment(docker, environment, containerId, instructionHandler);
+  }
+
+  private final Object lock = new Object();
+  private final DockerCommand docker;
+  private final Environment environment;
+  private final String containerId;
+  private final InstructionRequestHandler instructionHandler;
+
+  private DockerContainerEnvironment(
+      DockerCommand docker,
+      Environment environment,
+      String containerId,
+      InstructionRequestHandler instructionHandler) {
+    this.docker = docker;
+    this.environment = environment;
+    this.containerId = containerId;
+    this.instructionHandler = instructionHandler;
+  }
+
+  @Override
+  public Environment getEnvironment() {
+    return environment;
+  }
+
+  @Override
+  public InstructionRequestHandler getInstructionRequestHandler() {
+    return instructionHandler;
+  }
+
+  /**
+   * Closes this remote docker environment. The associated {@link InstructionRequestHandler} should
+   * not be used after calling this.
+   */
+  @Override
+  public void close() throws Exception {
+    synchronized (lock) {
```

Comment:
<!--new_thread; commit:21a3ef1d6e9deb78705f7e006ab32e103101f6df; resolved:0-->
Could you add a comment why we add this synchronization? Is it because calling instructionHandler.close() from two threads at the same time is unsafe, or calling docker.killContainer(containerId) is unsafe, or because it's unsafe if two threads both do instructionHandler.close() and then both do docker.killContainer()? (but all of this is safe if done from two threads sequentially?)